core/rio/turtle/src/main/java/org/eclipse/rdf4j/rio/turtle/TurtleWriter.java

```diff
@@ -129,6 +151,71 @@ public void endRDF()
 		}
 
 		try {
+			if (prettyPrintModel != null) {
+				// Note: We can't guarantee ordering at this point because Resource doesn't implement Comparable<Resource>
+				for (Resource nextContext : prettyPrintModel.contexts()) {
+					for (Resource nextSubject : prettyPrintModel.subjects()) {
+						boolean canShortenSubjectBNode = true;
+						// We can almost always shorten subject IRIs, 
+						// with some known corner cases that are already embedded in the algorithm
+						// So just need to do checking for BNode subjects
+						if (nextSubject instanceof BNode) {
+							if (prettyPrintModel.contains(null, null, nextSubject)) {
+								// Cannot shorten this blank node as it is used as the object of a statement somewhere 
+								// so must be written in a non-anonymous form
+								canShortenSubjectBNode = false;
+							}
```

Comment:
To be able to handle this kind of nested blank nodes (such as in anonymous objects, or list structures) we need to build up a dependency mapping of some sort. Or rather: if we have a case where the subject is the object of another statement, we can still shorten it if we _first_ print that other statement. 

Anyway, I realize there's more to it than that, so I'm fine with the implementation as-is for now. Just trying to think about future extension.